### PR TITLE
Enable clang sanitizers for C++ code and unit tests

### LINKED ISSUE #2978: Create Github-CI workflows that apply Clang sanitizers to the unit test execution
> estimate 2

Done criteria:
- [ ] Update github workflow to compile using two sanitization build work flows
- [ ] Figure out how to run the unit tests with the sanitizers engaged

### PR DESCRIPTION
**Pull request by avilcheslopez**
- Relates to #2978 
- Fixes #2978 #2509

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -45,6 +45,16 @@ jobs:
           cc: "gcc-9", cxx: "g++-9"
         }
         debug-flag: ["enable-debug", "disable-debug"]
+        service-sanitizers-flag: ["disable-sanitizers", "enable-sanitizers"]
+        base-sanitizers-flag: ["disable-sanitizers", "enable-sanitizers"]
```
Comment (cmcantalupo) [inline .github/workflows/build.yml]: maybe "base" -> "runtime" ?

File: service/configure.ac [hunk 2]
```diff
@@ -47,6 +47,20 @@ fi
 [enable_coverage="0"]
 )
 
+AC_ARG_ENABLE([sanitizers],
+  [AS_HELP_STRING([--enable-sanitizers], [Build and test with sanitizers enabled])],
```
Comment (cmcantalupo) [inline service/configure.ac]: I thought that there were two mutually exclusive sets of sanitizer options?  If there is more than one sanitizer configuration setting, we may need to be more specific, e.g. --enable-address-sanitizer / --enable-memory-sanitizer

File: service/configure.ac [hunk 3]
```diff
@@ -148,6 +162,11 @@ if test "x$enable_coverage" = "x1" ; then
 fi
 AC_SUBST([enable_coverage])
 
+if test "x$enable_sanitizers" = "x1" ; then
+  EXTRA_CXXFLAGS="$EXTRA_CXXFLAGS -fsanitize=address -mllvm -asan-use-private-alias=1"
+  #EXTRA_LDFLAGS="$EXTRA_LDFLAGS -fsanitize=address -mllvm -asan-use-private-alias=1"
```
Comment (cmcantalupo) [inline service/configure.ac]: we can remove this comment since it doesn't seem to be necessary with C++ code.

**Review by cmcantalupo — COMMENTED** [review]
Some small suggestions.  Looks good though!

**Comment by cmcantalupo** [comment]
```

Invalid workflow file: .github/workflows/build.yml#L53
The workflow is not valid. .github/workflows/build.yml (Line: 53, Col: 13): Matrix exclude key 'config.name' does not match any key within the matrix .github/workflows/build.yml (Line: 55, Col: 13): Matrix exclude key 'config.name' does not match any key within the matrix
```

File: .github/workflows/build.yml (hunk 1 above)
Comment (avilcheslopez) [reply .github/workflows/build.yml]: Discussed with Chris: we'll use the term "base" for now for consistency and make the change everywhere in a separate PR.

File: configure.ac [hunk 4]
```diff
@@ -188,6 +165,52 @@ fi
...
+  EXTRA_CXXFLAGS="$EXTRA_CXXFLAGS -fsanitize=address -fno-omit-frame-pointer -mllvm -asan-use-private-alias=1"
+fi
+
+if test "x$enable_overhead" = "x1" ; then
+  AC_DEFINE([GEOPM_OVERHEAD], [ ], [Enables code for measuring overhead])
+fi
+AC_SUBST([enable_overhead])
+
+AC_SUBST([enable_asan])
+AM_CONDITIONAL([ENABLE_SANITIZERS], [test "x$enable_asan" = "x1"])
```
Comment (cmcantalupo) [inline configure.ac:211]: maybe `ENABLE_SANITIZERS` -> `ENABLE_ASAN`

**Review by cmcantalupo — APPROVED** [review]
Thanks for doing all this clean up.  The Controller test in particular was a mess. Looks great.

**Pull request merged**